core/src/main/java/com/linecorp/armeria/client/AutoUpdatingAddressResolver.java

```diff
@@ -0,0 +1,380 @@
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.internal.dns.DefaultDnsNameResolver;
+import com.linecorp.armeria.internal.dns.DnsQuestionWithoutTrailingDot;
+
+import io.netty.channel.EventLoop;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsRecord;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.resolver.AbstractAddressResolver;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.Promise;
+
+class AutoUpdatingAddressResolver extends AbstractAddressResolver<InetSocketAddress> {
+
+    private static final Logger logger = LoggerFactory.getLogger(AutoUpdatingAddressResolver.class);
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<CacheEntry, ScheduledFuture> futureUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(CacheEntry.class, ScheduledFuture.class,
+                                                   "cacheUpdatingScheduledFuture");
+
+    /**
+     * A {@link ScheduledFuture} which is set in {@link CacheEntry} when the
+     * {@link AutoUpdatingAddressResolver} is closed by replacing and cancelling previously scheduled automatic
+     * DNS cache updating.
+     */
+    @VisibleForTesting
+    @SuppressWarnings("ComparableImplementedButEqualsNotOverridden")
+    static final ScheduledFuture<?> closed = new ScheduledFuture<Object>() {
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return false;
+        }
+
+        @Override
+        public long getDelay(TimeUnit unit) {
+            return Long.MIN_VALUE;
+        }
+
+        @Override
+        public int compareTo(Delayed o) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return true;
+        }
+
+        @Override
+        public boolean isDone() {
+            return true;
+        }
+
+        @Override
+        public Object get() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Object get(long timeout, TimeUnit unit) {
+            throw new UnsupportedOperationException();
+        }
+    };
+
+    private final ConcurrentMap<String, CompletableFuture<CacheEntry>> cache;
```

Comment:
Does this need to be concurrent? I think resolution happens on a single event loop.

Reply:
Yes. 😄 The cache is created in `AutoUpdatingAddressResolverGroup` and shared across from all `DnsResolver`s so that only one `EventLoop` sends the query for the hostname.